quinn-proto/src/connection/mod.rs

```diff
@@ -3963,6 +3963,23 @@ fn negotiate_max_idle_timeout(x: Option<VarInt>, y: Option<VarInt>) -> Option<Du
     }
 }
 
+/// A buffer that can tell how much has been written to it already
+///
+/// This is commonly used for when a buffer is passed and the user may not write past a
+/// given size.  It allows the user of such a buffer to know the current cursor position in
+/// the buffer.  The maximum write size is usually passed in the same unit as
+/// [`BufLen::len`]: bytes since the buffer start.
+pub(crate) trait BufLen {
```

Comment:
As suggested, can we make `BufLen: BufMut`?

Reply:
I chose not to do so because I don't think `BufLen` is that nice a trait.  It is more of a plug to bridge the use of `TransmitBuf` and the existing concept of passing around a buffer with a bunch of offsets describing it.  I hope that later it can be removed again, but also added this to limit the scope of this refactor.

And finally I think the naming would be much less clear, right now places get `BufMut + BufLen` and you immediately have very clear expectations.  If it were to be a supertrait it'd be something like `WriteBuffer`, `BufMutWithLen` or I don't know.  Naming it would become much trickier.

I do appreciate that this could become unwieldy as more traits are added.  Though all the uses for now are just two traits, which I don't think is problematic yet.